src/edu/washington/escience/myria/operator/SamplingDistribution.java

```diff
@@ -0,0 +1,320 @@
+    this.randomSeed = randomSeed;
+  }
+
+  /**
+   * Instantiate a SamplingDistribution operator using a specific sample size.
+   *
+   * @param sampleSize
+   *          total samples to create a distribution for.
+   * @param sampleType
+   *          the type of sampling distribution to create
+   * @param child
+   *          extracts (WorkerID, PartitionSize, StreamSize) information from
+   *          this child.
+   * @param randomSeed
+   *          value to seed the random generator with. null if no specified seed
+   */
+  public SamplingDistribution(Operator child, int sampleSize,
+      SamplingType sampleType, Long randomSeed) {
+    this(child, sampleType, randomSeed);
+    this.sampleSize = sampleSize;
+    Preconditions.checkState(this.sampleSize >= 0,
+        "Sample Size must be >= 0: %s", this.sampleSize);
+  }
+
+  /**
+   * Instantiate a SamplingDistribution operator using a percentage of total
+   * tuples.
+   *
+   * @param samplePercentage
+   *          percentage of total samples to create a distribution for.
+   * @param sampleType
+   *          the type of sampling distribution to create
+   * @param child
+   *          extracts (WorkerID, PartitionSize, StreamSize) information from
+   *          this child.
+   * @param randomSeed
+   *          value to seed the random generator with. null if no specified seed
+   */
+  public SamplingDistribution(Operator child, float samplePercentage,
+      SamplingType sampleType, Long randomSeed) {
+    this(child, sampleType, randomSeed);
+    this.isPercentageSample = true;
+    this.samplePercentage = samplePercentage;
+    Preconditions.checkState(samplePercentage >= 0 && samplePercentage <= 100,
+        "Sample Percentage must be >= 0 && <= 100: %s", samplePercentage);
+  }
+
+  @Override
+  protected TupleBatch fetchNextReady() throws DbException {
+    // Drain out all the workerID and partitionSize info.
+    while (!getChild().eos()) {
+      TupleBatch tb = getChild().nextReady();
+      if (tb == null) {
+        if (getChild().eos()) {
+          break;
+        }
+        return null;
+      }
+      Type col0Type = tb.getSchema().getColumnType(0);
+      Type col1Type = tb.getSchema().getColumnType(1);
+      boolean hasStreamSize = false;
+      Type col2Type = null;
+      if (tb.getSchema().numColumns() > 2) {
+        hasStreamSize = true;
+        col2Type = tb.getSchema().getColumnType(2);
+      }
+      for (int i = 0; i < tb.numTuples(); i++) {
+        int workerID;
+        if (col0Type == Type.INT_TYPE) {
+          workerID = tb.getInt(0, i);
+        } else if (col0Type == Type.LONG_TYPE) {
+          workerID = (int) tb.getLong(0, i);
+        } else {
+          throw new DbException("WorkerID must be of type INT or LONG");
+        }
+        Preconditions.checkState(workerID > 0, "WorkerID must be > 0");
+        // Ensure the future .set(workerID, -) calls will work.
+        for (int j = tupleCounts.size(); j < workerID; j++) {
```

Comment:
Are `workerID`s guaranteed to be consecutive and always begin at 1? If not (and I don't think they will be, since relations are usually not partitioned on all workers), we should use a `Map` instead for `tupleCounts`/`streamCounts`, otherwise you'll have "dummy" entries which could cause bugs.